csrc/scheduler/mma_utils.cpp

```diff
@@ -1575,37 +1574,62 @@ RolesMapOpt getTensorsRoles(Fusion* fusion) {
       bool has_n = (end != std::find(begin, end, MatmulDomain::N));
       bool has_k = (end != std::find(begin, end, MatmulDomain::K));
 
-      if (!processing_output && has_m && has_k && !has_n) {
+      if (has_m && has_k && !has_n) {
         roles_map[MatmulRole::INPUT_A].push_back(entry.first);
         continue;
       }
-      if (!processing_output && has_n && has_k && !has_m) {
+      if (has_n && has_k && !has_m) {
         roles_map[MatmulRole::INPUT_B].push_back(entry.first);
         continue;
       }
-      if (!processing_output && has_m && has_n && !has_k) {
+      if (has_m && has_n && !has_k) {
         roles_map[MatmulRole::INPUT_C].push_back(entry.first);
         continue;
       }
       // Bias vectors are assigned to INPUT_C role
-      if (!processing_output && has_m && !has_n && !has_k) {
+      if (has_m && !has_n && !has_k) {
         roles_map[MatmulRole::INPUT_C].push_back(entry.first);
         continue;
       }
+    }
+    for (auto& [role, tvs] : roles_map) {
+      // sort tvs by name()
+      std::sort(tvs.begin(), tvs.end(), [](TensorView* a, TensorView* b) {
+        return a->name() < b->name();
+      });
+    }
+  };
+
+  const auto findOutputRolesByDomains = [](const DependenciesMap& deps_map,
+                                           RolesMap& roles_map) {
+    for (const auto& entry : deps_map) {
+      const auto& domains = entry.second;
+      const auto begin = domains.begin();
+      const auto end = domains.end();
+
+      bool has_m = (end != std::find(begin, end, MatmulDomain::M));
+      bool has_n = (end != std::find(begin, end, MatmulDomain::N));
 
       // NOTE: depending on fusion definition k domain may appear in the output:
       //  - for mma_output == fusion output k domain is present
       //  - for mma_output != fusion output (fusion with epilogue) k domain
       //    is not present
-      if (processing_output && has_m && has_n) {
+
+      // NOTE: the core fusion output tensors are the ones with m and n
+      // domains
+      if (has_m && has_n) {
```

Comment:
With proposed implementation (use of outputs' cache and fork) the role of `OUTPUT_D` is reduced to the point it may not be needed.

I wonder if it is needed, or we can change the analysis to pick a single `OUTPUT_D` tv with `m` and `n` with the largest number of uses from all outputs? The rest of output tvs would be assigned to `OUTPUT_AUX` role - this would be required to check if all input/output tvs have a role assigned.

For simple matmul it will be the only output.
For fusion with multiple outputs the 'reference' output will be tv used to produce other outputs. For fusion with three outputs:
- `matmul`
- `matmul + bias`
- `matmul + bias + relu`

the first tv (`matmul`) would be `OUTPUT_D` tensor. This might be useful for some analysis.

Reply:
I think for the near future, (matmul, matmul + bias, matmul + bias + activation), these two approaches have no difference. For longer future, if we want to supported general fusions with complicated broadcasting patterns, then yes, we do need a reference tensor, but the reference tensor should not be the one with most uses, but instead, it should be the one that contains the most information about `IterDomain`s. Regarding this PR, I have no preference about which approach we take. Probably whichever is easier is better.

Reply:
Ok, it makes sense. To prepare matmul scheduler for a long term direction I will prepare then the following roles:
- `OUTPUT_D`, which will be our reference,
- `OUTPUT_AUX`, which will be for all matmul fusion outputs with exception for the `OUTPUT_D`

For the time being, the expectation is that `OUTPUT_D` and `OUTPUT_AUX` will have `m` and `n` IterDomains. 

The reason for this split of roles is to aid analysis in controlling if all input/output tensors have assigned roles.

In future we can change conditions for assigning `OUTPUT_D` role so it will point to a reference tensor.